Added updateUser call

Merseleo opened the pull request:
Closes #31 

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a service, which does x -->

- I added the implementation for the following call
  - updateUser
- This includes the implementation of the repository layer, i.e., an update function for a user (probably need to be generalized when introducing #93), the service layer and the validation for an update request.
- The validation especially includes a check, whether a field mask is valid
- Furthermore I figured out how to manually test this call in IntelliJ: 
![Bildschirmfoto 2025-07-03 um 19 46 09](https://github.com/user-attachments/assets/addd9bd3-39b9-4cd4-8573-baa404c05768)
but this doesn't work as long as you cannot login using manual requests
- Regarding to the issue #31 this PR should also implement the `ChangePassword` call, however the overlap of these calls is not as large as expected, so this call should be implemented separately (see #93).

>**Note**: I do not understand why the coverage reporter does not recognize the coverage for the `UserTableRepo`. When I run the tests locally and check the coverage, my new implementation is completely covered. I am glad for suggestions and ideas for improvement as this is my first PR for the backend.

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

Commit:
feat: add repository implementation for updateUser

Furthermore, the UserTableRepo and its corresponding tests were slightly refactored to reduce duplicate code.

Commit:
feat: add Unauthorized exception for the case that a user want to access another single user

Accessing another user means creating, reading, updating or deleting this user. Possible access operations are specified in the AccessType.kt.

Commit:
feat: add service implementation for updateUser

mowi12 committed:
chore: move GrpcContext from grpc package to auth

The layered architecture tests prohibit the access of the grpc layer from the service layer. Therefore, the GrpcContext needs to be moved out of the grpc layer into an unrestricted layer.

Commit:
feat: add validation for updateUser request

Slartibartfass2 commented on build.gradle.kts:
```diff
@@ -67,6 +67,8 @@ dependencies {
 
     detektPlugins(libs.detekt.formatting)
+
+    implementation("com.google.protobuf:protobuf-java-util:3.19.4")
```
Please declare this in the same way as the other dependencies:
- This line can go to line 66 where other protobuf stuff is declared
- Add `protobuf-java-util = { module = "com.google.protobuf:protobuf-java-util", version.ref = "protobuf-kotlin-version" }` in `gradle/libs.versions.toml` in line 74
- Then you can change the previously added line to `implementation(libs.protobuf.java.util)`

Note that you have to update the gradle project, either by clicking the symbol in the upper-right corner or by using the shortcut Ctrl+Shift+O

Merseleo replied:
Ah thank you. Did not understand how to properly add dependencies, a very useful tip.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/model/ValidationIssue.kt:
```diff
@@ -115,3 +115,21 @@ data class InvalidPassword(
 }
+
+/**
+ * Represents an issue where a provided field mask contains invalid fields or is blank.
```
For consistency.
```suggestion
 * Represents a validation issue where a provided field mask contains invalid fields or is blank.
```

Merseleo replied:
You are right, I missed it.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -67,13 +87,22 @@ interface IUserTableRepo {
             .map { it.toUser() }
             .singleOrNull()
-            ?: throw NotFoundException.User(id.toString())
+    }
```
The "suspend" is not necessary. What about this name? That it's from the database is obvious because we're in the repository layer.
```suggestion
    private fun getUserByIdOrNull(id: UUID): User? =
        UserTable
            .selectAll()
            .where { UserTable.id eq id }
            .map { it.toUser() }
            .singleOrNull()
```

Merseleo replied:
Yeh I wanted to differentiate this private function from the `getUserById` function, but probably your wording is better.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -113,11 +142,33 @@ class UserTableRepo(
+        if (updatedRowCount == 0) {
+            // If no row was updated, then the user with the given id does not exist
+            throw NotFoundException.User(uuid.toString())
+        }
```
Dominik and I already discussed this in #60 and came to the conclusion that we don't need to test whether the user exists again because we already do this on the service layer by calling `getUserById`. What is your opinion on that?

Merseleo replied:
I thought we should implement this layer without any assumptions about the layer above? So what I want to say: now we only use this repository layer function by this one service layer function that already check that the user exist (besides, theoretically it is possible, that this user doesn't exist anymore when this repository layer function is executed as the service layer function isn't atomic), but isn't it possible that this repository layer function is used by another service layer function that doesn't check this? So I do not really agree with you.

Merseleo replied:
However, I see your point and we can also always assume that the layer above only pass valid objects down and handle everything.

Slartibartfass2 replied:
I see your point, but then we need to make clear assumptions about our layers because so far, we didn't. This also introduces some duplicate code because then we need to check whether an entity exists on both layers. Maybe we should discuss this with the group.

Merseleo replied:
Duplicate is not the correct wording as this code wouldn't be duplicate, right? But I agree that in some cases we would do double checks. I will remove it for now to keep our code consistent, but let's discuss these things on Thursday.

Slartibartfass2 replied:
Well, it's double-checking the same thing. Also, the user object always exists even after deletion, so we won't experience race conditions with that.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -113,11 +142,33 @@ class UserTableRepo(
+                    "user.email" -> it[email] = request.user.email
+                    "user.first_name" -> it[firstName] = request.user.firstName
+                    "user.last_name" -> it[lastName] = request.user.lastName
+                    "user.role" -> it[role] = request.user.role
```
What about the status and the password hash?

Merseleo replied:
I only wanted to implement everything in this PR that can actually can be updated with the `updateUser` call. I agree, status and password hash can also be updated, but not with the help of this FieldMask object, but via separate gRPCs. Accordingly, I would like to adapt this repository method when we implement these calls. I think we should only implement as much as is necessary for the feature. I was already thinking about this yesterday and when these updates come you might have to change the signature of the method and then the comments and calls etc. anyway, but in my opinion that doesn't belong here.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -113,11 +142,33 @@ class UserTableRepo(
+                    "user.first_name" -> it[firstName] = request.user.firstName
+                    "user.last_name" -> it[lastName] = request.user.lastName
+                    "user.role" -> it[role] = request.user.role
+                    else -> throw IllegalArgumentException("$field is not a valid field mask entry for the user.")
```
Don't we validate the field mask in the validation layer?

Merseleo replied:
Yes, we do. This was something I wasn't sure about. Do you think, this check is unnecessary and can be deleted?

Slartibartfass2 replied:
Yes, if think it is okay to make this assumption. Also, nothing happens if the field mask is empty, so no harm there.

Merseleo replied:
Ok, then I delete it.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -148,4 +161,27 @@ class UserService(
     }
+
+    override suspend fun updateUser(request: UserOuterClass.User.Update): UserOuterClass.User {
+        val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
```
By adding this, we also check whether the requested user even exists, which is currently missing here.
```suggestion
        val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
        val requestedUserId = parseUUID(request.user.id, "user")
        val requestedUser = userRepo.getUserById(requestedUserId)
```

Merseleo replied:
You are right, this test of existence I outsourced to the repository layer by checking whether the update worked. I am not sure, where it does belong and whether we really need to check this here?

Slartibartfass2 replied:
This should be resolved in the discussion above. But anyway, I thought about it again and in general, we should first make the access check and then any other check. So, please move the access check before the one for the duplicate email.

Merseleo replied:
I added your suggestion. Did I miss anything as I moved the checks around?

Slartibartfass2 replied:
Looks good.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -148,4 +161,27 @@ class UserService(
+            throw DuplicateEntityException.UserEmail(request.user.email)
+        }
+
+        // When the role should be changed, then check whether the requesting user a server admin
```
```suggestion
        // When the role should be changed, then check whether the requesting user is a server admin
```

Merseleo replied:
Thank you

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -148,4 +161,27 @@ class UserService(
+        // When the role should be changed, then check whether the requesting user a server admin
+        if (request.mask.pathsList.contains("role") && currentUser.role != UserRole.USER_ROLE_ADMIN) {
+            throw UnauthorizedException.Single.User(currentUser.id.toString(), request.user.id, AccessType.UPDATE)
+        }
```
Meanwhile, this helper was added.
```suggestion
        if (request.mask.pathsList.contains("role")) {
            verifyServerAdminRole(currentUser) { UnauthorizedException.Single.User(it, request.user.id, AccessType.UPDATE) }
        }
```

Merseleo replied:
Yeah thank you for the hint, but this does not work directly, as this method only accept an `UnauthorizedException.All`

Slartibartfass2 replied:
Yes, we have to find a way to generalize this after you rebase.

Merseleo replied:
Ok, for now I changed the signature to accept all kinds of Unauthorized Exceptions.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -148,4 +161,27 @@ class UserService(
+        // its own information or changes to information from another user (except the user is an admin)
+        if (currentUser.id.toString() != request.user.id && currentUser.role != UserRole.USER_ROLE_ADMIN) {
+            throw UnauthorizedException.Single.User(currentUser.id.toString(), request.user.id, AccessType.UPDATE)
+        }
```
Here you can use the helper too. But you can also combine both checks, as they do the same.

```suggestion
        // Check whether the current user is a server admin if the role is changed or the requested user is different
        // from the current user
        if (request.mask.pathsList.contains("role") || currentUser.id != requestedUser.id) {
            verifyServerAdminRole(currentUser) { UnauthorizedException.Single.User(it, request.user.id, AccessType.UPDATE) }
        }
```

Merseleo replied:
I combined them.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/model/SnowballRException.kt:
```diff
@@ -102,28 +102,37 @@ sealed class SnowballRException(
      * type and ID of the accessed entity.
      * @param currentUserId The ID of the user that is accessing the entity.
      * @param accessedEntityMessage The message of what is accessed.
+     * @param accessType The type of access, i.e., an update, read access, ...
```
This is a great addition to the parameterization done in #130.

Merseleo replied:
But this parameterization will work together with my change without any problems, right?

Slartibartfass2 replied:
yes, only merge conflicts :/

Merseleo replied:
Ok 👍

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/validation/UserValidator.kt:
```diff
@@ -0,0 +1,54 @@
+ */
+object UserValidator {
+    fun validateUpdateRequest(request: User.Update): EitherNel<ValidationIssue, Unit> = either {
+        // validate the field mask
```
```suggestion
        // Validate the field mask
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/validation/ValidationHelper.kt:
```diff
@@ -141,3 +145,18 @@ fun Raise<ValidationIssue>.ensureLastNameValidity(lastName: String) {
+ * Ensures that the provided field mask contains only valid fields for the given object type and
+ * is non-blank.
+ *
+ * @param fieldMask The FieldMask to validate.
```
```suggestion
 * @param fieldMask The [FieldMask] to validate.
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/validation/ValidationHelper.kt:
```diff
@@ -141,3 +145,18 @@ fun Raise<ValidationIssue>.ensureLastNameValidity(lastName: String) {
+ * @param fieldMask The FieldMask to validate.
+ * @param descriptor The object descriptor to validate against.
+ * @throws ValidationIssue if the field mask contains invalid or blank fields.
+ */
```
We don't throw anything here.
```suggestion
 * @param descriptor The object descriptor to validate against.
 */
```

Merseleo replied:
You are right. Initially, I thought that this raise type was just a fancier exception, but now I understand better what it is.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/validation/UserValidator.kt:
```diff
@@ -0,0 +1,54 @@
+            },
+            {
+                if ("user.role" in selectedFields) {
+                    ensureEnumNotUnspecified("userRole", request.user.role)
```
The name of the field is just `role`.

```suggestion
                    ensureEnumNotUnspecified("role", request.user.role)
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/UserTableRepoTest.kt:
```diff
@@ -10,44 +11,67 @@ import org.junit.jupiter.api.Assertions.assertTrue
+    companion object {
+        @JvmStatic
+        fun validFieldMasks(): Stream<Arguments> = Stream.of(
+            Arguments.of(listOf("user.email")),
+            Arguments.of(listOf("user.first_name", "user.last_name")),
+            Arguments.of(listOf("user.role")),
```
```suggestion
            Arguments.of("user.email"),
            Arguments.of("user.first_name", "user.last_name"),
            Arguments.of("user.role"),
```

Merseleo replied:
I think this cannot be deleted, as the arguments need to be lists as I build the field mask from a list of strings. When I delete this `listOf(...)` it doesn't work anymore to build the field mask from a string list.

Slartibartfass2 replied:
Ah yes, I got that wrong.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/UserTableRepoTest.kt:
```diff
@@ -10,44 +11,67 @@ import org.junit.jupiter.api.Assertions.assertTrue
+
+    companion object {
+        @JvmStatic
+        fun validFieldMasks(): Stream<Arguments> = Stream.of(
```
only FYI: this doesn't have to be a stream, you could simply use a list.

Merseleo replied:
Thank you for the hint.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/UserTableRepoTest.kt:
```diff
@@ -214,4 +194,110 @@ class UserTableRepoTest : H2DatabaseTest(arrayOf(UserTable)) {
+            if ("user.last_name" in fieldMask) updatedUserBuilder.lastName = "Doe"
+            if ("user.role" in fieldMask) updatedUserBuilder.role = UserRole.USER_ROLE_ADMIN
+
+            val request = UserOuterClass.User.Update.newBuilder()
```
```suggestion
            val request = User.Update.newBuilder()
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/UserTableRepoTest.kt:
```diff
@@ -214,4 +194,110 @@ class UserTableRepoTest : H2DatabaseTest(arrayOf(UserTable)) {
+            if ("user.email" in fieldMask) updatedUserBuilder.email = "updated.user@example.com"
+            if ("user.first_name" in fieldMask) updatedUserBuilder.firstName = "John"
+            if ("user.last_name" in fieldMask) updatedUserBuilder.lastName = "Doe"
+            if ("user.role" in fieldMask) updatedUserBuilder.role = UserRole.USER_ROLE_ADMIN
```
Do we need the if statements? If we would just set the fields, it would alright too, wouldn't it? I think it would be beneficial because then we would see whether changed fields that are not in the field mask are applied.

Merseleo replied:
You are right, actually it is even better to always you the same update object and check whether only these fields were updated that were specified in the field mask, good catch!

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/UserTableRepoTest.kt:
```diff
@@ -214,4 +194,110 @@ class UserTableRepoTest : H2DatabaseTest(arrayOf(UserTable)) {
+                )
+            } else {
+                assertThat(updatedUser.role).isEqualTo(UserRole.USER_ROLE_DEFAULT)
+            }
```
These seem unnecessarily wrapped.

```suggestion
            if ("user.email" in fieldMask) {
                assertThat(updatedUser.email).isEqualTo("updated.user@example.com")
            } else {
                assertThat(updatedUser.email).isEqualTo("test.user@example.com")
            }
            if ("user.first_name" in fieldMask) {
                assertThat(updatedUser.firstName).isEqualTo("John")
            } else {
                assertThat(updatedUser.firstName).isEqualTo("Test")
            }
            if ("user.last_name" in fieldMask) {
                assertThat(updatedUser.lastName).isEqualTo("Doe")
            } else {
                assertThat(updatedUser.lastName).isEqualTo("User")
            }
            if ("user.role" in fieldMask) {
                assertThat(updatedUser.role).isEqualTo(UserRole.USER_ROLE_ADMIN)
            } else {
                assertThat(updatedUser.role).isEqualTo(UserRole.USER_ROLE_DEFAULT)
            }
```

Merseleo replied:
You are right, this was automatically formatted by the formatter, I do not quite understand why.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/UpdateTest.kt:
```diff
@@ -0,0 +1,89 @@
+
+@ExperimentalCoroutinesApi
+@DelicateCoroutinesApi
+internal class UpdateTest : MainServiceTest() {
```
So far, we named the file and class according to the service method, so let's keep it consistent and name this `UpdateUserTest`.

Slartibartfass2 replied:
Also, ever mock every external call in the server method so that we can ensure that we took the correct path in our test case. In this file this would be `userRepo.getUserById`, `userRepo.doesUserExistByEmail`, and `userRepo.updateUser`.

Merseleo replied:
My mistake, sorry. I also deleted the internal keyword.

Merseleo replied:
But I think I do not understand your second comment.

Slartibartfass2 replied:
Take for example the first test, in there we mock `getUserById`, but `userRepo.updateUser(request)` is executed as well in this service method, and we do not mock its behavior. This is fine because the mocking library just does nothing by default, but this is the general question we have to ask ourselves if we shouldn't mock all calls no matter if their behavior is important for the method or not. With this, we could also make sure that some paths are not taken. e.g. by mocking that `userRepo.doesUserExistByEmail(request.user.email)` throws an exception, we can ensure that this isn't tested in this test case.

I'm quite unsure how to deal with this, but we have to find a consistent way to test our methods.

Maybe we could do something like this:
```kotlin
    @BeforeEach
    override fun setUpTest() {
        super.setUpTest()

        coEvery { userRepoMock.getUserById(any()) } throws NotImplementedError()
        coEvery { userRepoMock.doesUserExistByEmail(any()) } throws NotImplementedError()
        coEvery { userRepoMock.updateUser(any()) } throws NotImplementedError()
    }
```
This way, each test method would throw the `NotImplementedError` when a method is not mocked but called.

What are your thoughts on this?

Merseleo replied:
I think your suggestion is nice to figure out which calls should be mocked, but should not be kept after the implementation of the tests. But I will add all necessary mocks.

Slartibartfass2 replied:
Why not keep it? This would fail if the tests aren't adapted after the code has changed. It would be a hint that we have to update our tests.

Merseleo replied:
Also true again, I add them again.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/UpdateTest.kt:
```diff
@@ -0,0 +1,89 @@
+        coEvery { userRepoMock.getUserById(any()) } returns nonAdminUser
+
+        assertThrows<SnowballRException.UnauthorizedException.Single.User> { mainService.updateUser(request) }
+    }
```
Let's also add the two test cases where the user is an admin and the "role" or "email" field is in the field mask. Otherwise, we wouldn't know if the exception is only thrown because of the value in the field mask.

Merseleo replied:
I do not quite understand how these test cases should look like. Is the user to update an admin? Or do you want to extend the positive case, where the update of a normal user works?

Slartibartfass2 replied:
Not sure what I meant with the first test, but the second and the one below. Here we test if a non admin user either changes their role or the information of another use. For both cases we could add a test case where the user is an admin instead. This way we ensure that the `UnauthorizedException` is not only thrown because of the field masks having "role" in it or the current user accessing another user.

TL;DR test all branches

Merseleo replied:
Except for role I already test this in the test "When an admin updates a user's information successfully, then no exception is thrown"? There an admin changes the details of another non-admin user. Do you want me to extend this test case to include the check for the role? Because actually we test all branches, otherwise the branch coverage wouldn't be 100% or?

Slartibartfass2 replied:
We don't check it for this test case in the `verifyServerAdmin` method. If you would just add it to the existing test, this is fine.

Merseleo replied:
I am not sure whether I correctly understand what to add, but I added another check to my mentioned test. Was this correct?

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/UpdateTest.kt:
```diff
@@ -0,0 +1,89 @@
+
+        coEvery { userRepoMock.updateUser(any()) } throws Exception("Failed to persist user")
+
+        assertThrows<Exception> { mainService.updateUser(request) }
```
Use the `TestSpecificException` so that we can assure that the test case fails because of this mocked exception.
```suggestion
        coEvery { userRepoMock.updateUser(any()) } throws TestSpecificException()

        assertThrows<TestSpecificException> { mainService.updateUser(request) }
```

Merseleo replied:
This I new, right?

Slartibartfass2 replied:
yes

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/validation/UserValidatorTest.kt:
```diff
@@ -0,0 +1,176 @@
+        }
+
+        @Test
+        fun `When an invalid id is validated, then the 'InvalidId' issue is returned`() {
```
```suggestion
        fun `When an invalid ID is validated, then the 'InvalidId' issue is returned`() {
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/validation/UserValidatorTest.kt:
```diff
@@ -0,0 +1,176 @@
+        }
+
+        @Test
+        fun `When an invalid role is provided and specified in the field mask, then the 'InvalidEmail' issue is returned`() {
```
```suggestion
        fun `When an invalid role is provided and specified in the field mask, then the 'EnumUnspecified' issue is returned`() {
```

Merseleo replied:
A rest from copy & paste, thank you

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/model/AccessType.kt:
```diff
@@ -0,0 +1,21 @@
+ */
+enum class AccessType(val description: String) {
+    /** Attempt to read or access an entity */
+    READ("access"),
```
What's the difference between read and access?

Merseleo replied:
There is no difference, in my opinion the wording "access" only sounds better in the error message than "read", but I can also change it to "read". However, I wanted to name the AccessType after the CRUD operations that are possible. 
What do you think? Should be write "read" in the error message for consistency?

Slartibartfass2 replied:
I think the description of `READ` should be `read` similar to the other values. Access is very ambiguous and also the same name as the whole enum.

Merseleo replied:
Ok, i will rename it.

Slartibartfass2 commented on detekt.yml:
```diff
@@ -94,7 +94,7 @@ complexity:
     CognitiveComplexMethod:
         active: true
-        threshold: 15
+        threshold: 25
```
What about excluding all test files with `excludes: [ '**/test/**' ]` or only ignoring this single method instead? I think this metric is important for the production code.

Merseleo replied:
You are right, probably your solution is better.

Slartibartfass2 reviewed (changes requested):
Great job! For your first PR in this repo, it is remarkable how well you implemented the different layers. And you also figured out how to deal with the field masks, which is significant for other update calls.

The decrease in coverage in the user table repo is because of the `EntityNotPersistedException`, but this is fine.

Merseleo committed:
docs: improve comments and test descriptions

Typos, bad wording or missing references in the comments or test descriptions were removed.

Co-authored-by: Felix Schlegel <40503329+Slartibartfass2@users.noreply.github.com>

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -113,11 +142,33 @@ class UserTableRepo(
+                }
+            }
+
+            it[modifiedAt] = OffsetDateTime.now()
```
And something I forgot to ask in the review myself: Now I always update the `modifiedAt` date even, when the updated value is the same and so no user details change. Is this behavior we want or should we always check, whether we really _modified_ anything?

Merseleo replied:
@Slartibartfass2 What do you think?

Slartibartfass2 replied:
Hard to tell, we could make the assumption that the field mask is not empty or we just don't care whether this is updated or not. Not sure if you can test whether an update really applied. I can't test this right now.

Merseleo replied:
Yeah in my opinion it is a valid assumption that the field mask isn't empty. But then let's keep it as it is for now.

Merseleo reviewed (commented):
Thanks for the detailed review, there were some useful tips. However, I didn't quite understand some of the comments or their meaning, perhaps we can discuss this in more detail again. But with the call then many of the other calls should be really straight forward (except for the whole paper fetching and reviewing thing).

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/UpdateUserTest.kt:
```diff
@@ -0,0 +1,96 @@
+@DelicateCoroutinesApi
+class UpdateUserTest : MainServiceTest() {
+    @Test
+    fun `When a user updates its own information successfully, then no exception is thrown`() = testCoroutine {
```
```suggestion
    fun `When a user updates their own information successfully, then no exception is thrown`() = testCoroutine {
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/user/UpdateUserTest.kt:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    @Test
+    fun `When a user wants to change its email to an existing email, then an exception is thrown`() = testCoroutine {
```
```suggestion
    fun `When a user wants to change their email to an existing email, then an exception is thrown`() = testCoroutine {
```

Slartibartfass2 reviewed (changes requested):
Great work. Let's finish the discussions and rebase after the next round.

Slartibartfass2 reviewed (changes requested):
misclicked

Commit:
fix: move the check whether the user that should be updated exists from the repository into the service layer

Now the repository layer assumes it gets a valid field mask and a valid user object.

Commit:
test: add missing mocks of functions called in the updateUser function in the service layer

Merseleo reviewed (commented):
I am still not quite happy with the "new experience" for reviewing: I hope I did not miss any of your comments. Anyway, now every is clear to me and I hope I changed the code correctly.

Slartibartfass2 reviewed (commented):
Good job, let's rebase.

Commit:
test: add a missing test case that an admin user can successfully change the role of another user

Merseleo reviewed (commented):
So I hope I fixed everything. As you requested I rebased this branch and hopefully did not introduce any errors. I also squashed the old commits used for easier reviewing, so I have now the final commit history.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/UserTableRepo.kt:
```diff
@@ -54,6 +60,20 @@ interface IUserTableRepo {
+     * - last name
+     * - email
+     * - role
+     * - password hash
```
```suggestion
     * - role
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -115,6 +123,7 @@ class UserService(
         val requestedUser = userRepo.getUserByEmail(request.email)
 
+        @Suppress("StringLiteralDuplication")
         verifyUserAccess(currentUser, requestingUserId, IdentifierType.EMAIL)
```
```suggestion
        verifyUserAccess(currentUser, requestingUserId, IdentifierType.EMAIL)
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/model/SnowballRException.kt:
```diff
@@ -58,41 +58,48 @@ sealed class SnowballRException(
          *
          * @param accessedEntityType The type of the entity that was accessed without permission.
          * @param accessedEntityId The ID of the entity that was accessed without permission.
+         * @param accessType The type of access (@see [AccessType] for possible types).
```
The `@see` annotation doesn't work that way, but you can just say 'see'

```suggestion
         * @param accessType The type of access (see [AccessType] for possible types).
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/model/SnowballRException.kt:
```diff
@@ -58,41 +58,48 @@ sealed class SnowballRException(
          * permission.
          *
          * @param accessedEntityType The type of the entities that were accessed without permission.
+         * @param accessType The type of access (@see [AccessType] for possible types).
```
```suggestion
         * @param accessType The type of access (see [AccessType] for possible types).
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -155,4 +164,35 @@ class UserService(
+                    AccessType.UPDATE,
+                    currentUser.id.toString(),
+                )
+            }
```
You can use `it` for the last param.
```suggestion
            verifyServerAdminRole(currentUser) {
                UnauthorizedException.Single(EntityType.USER, requestedUser.id.toString(), AccessType.UPDATE, it)
            }
```

I'm not happy with the way we treat this now, but I don't see another solution, so we keep this.

Slartibartfass2 reviewed (changes requested):
Only these few suggestions left.

Merseleo committed:
refactor: minor code and comment improvements

Co-authored-by: Felix Schlegel <40503329+Slartibartfass2@users.noreply.github.com>

Slartibartfass2 reviewed (approved):
Great job Leonhard.
